### DISCUSSION TITLE
I have two models (events & artists) pagy is working fine with my artist model, but my event model isn't displaying last records

### DISCUSSION THREAD

**Discussion by ogarocious** [ddnexus/pagy#450] [Q&A] [answered]
I do have a more elaborate index method in my events controller

```
def index
    if params[:q]
      params[:q][:combinator] = "and"
      params[:q][:groupings] = []
      split_geo = params[:q][:address_or_city_or_state_or_country_or_continent_cont_all].split((/(,\s*)+/))
      split_geo.map! do |a|
        I18n.transliterate a
      end
      split_geo.each_with_index do |word, index|
        params[:q][:groupings][index] = { address_or_city_or_state_or_country_or_continent_cont_all: word }
      end
    end
    @q = Event.ransack(params[:q])
    @pagy, @events = pagy(@q.result(distinct: true))
  end
```

In my events index page I have:

```
<h2>Upcoming Events</h2>
         <br>
         <%== pagy_bootstrap_nav(@pagy) %>
         <br>
              <div class="event-list-wrapper">
                  <% @events.upcoming_events.each do |event| %>
                    <%= render 'event', event: event %>
                  <% end %>
              </div>
```

When I remove my upcoming_events scope, the records will display correctly.

In my event.rb model I have:

` scope :upcoming_events, -> { where('event_start_date > ?', Time.now).or(where('event_end_date > ?', Date.yesterday)) }`

Am I missing something in order to get pagy with work with my event scope?

**Comment by benkoshy** [comment]
My best guess is that you ought to apply all scopes BEFORE passing into pagy:.

I"m not sure about the ransack operations required to make it work, but as far as pagy is concerned, I can suggest the following:

```diff
- @pagy, @events = pagy(@q.result(distinct: true))
+ @pagy, @events = pagy( get_upcoming_events_here)
```

```diff
- <% @events.upcoming_events.each do |event| %> #no don't apply after paginating, do it before
+ <% @events.each do |event| %>
```

**Reply by ogarocious** [reply]
Thanks for the reply! The reason I have the upcoming events scope in my view is because I have a section on this page for upcoming events and another for past events, so I'm using the scope to separate the by that time sorting of the scopes.

So in my index method for my events, the index is sent to display all the results based on any ransack searching and then from there the event time scopes come into play.

Any other ideas?

**Reply by benkoshy** [reply]
> So in my index method for my events, the index is sent to display all the results based on any ransack searching and then from there the event time scopes come into play.
> Any other ideas?

In that case, since this has little to do with scopes, why not use a view helper?

```rb
module EventHelper
   def is_upcoming?(event)
     return event.event_start_date > Time.now ||  event.event_end_date > Time.now.yesterday # untested
     # also might x2 check your logic. not sure why event dates which end yesterday onwards are still seen as upcoming?   
   end
end
```

Or you could otherwise use a model helper, but I would favour a view helper (FWIW).

```erb
<% @events.each do |event| %>
     <% if upcoming(event) %>
          <%= render 'event', event: event %>
    <% end %>    
<% end %>
```

**Comment by ddnexus** [comment]
> The reason I have the upcoming events scope in my view is because I have a section on this page for upcoming events and another for past events, so I'm using the scope to separate the by that time sorting of the scopes.
... the index is sent to display all the results based on any ransack searching and then from there the event time scopes come into play

Not very clear what you describe, but if you mean that you have 2 sections in the view, you should use 2 different sets of events and pagy objects that work independently from each other.

**Reply by ogarocious** [reply]
![image](https://user-images.githubusercontent.com/91164935/214639195-5460f2f8-59c5-434d-9613-36007623c491.png)

This is an image that shows what's happening. When you say to separate sets of events, how should that separation happen?

**Reply by benkoshy** [reply]
Sounds like you are trying to mix future and past events. I wouldn't do it the way you're attempting above.  I think what you want is to separate past events and future events, and to paginate both separately. What (I believe) Domizio was alluding to was something like this:

```rb
@events_pagy, @events = pagy(events_scope)
@upcoming_pagy, @upcoming = pagy(upcoming, page_param: :upcoming_page) # ddnexus is correct you need a different page param. sorry i missed this.
```

```erb
<% @events.each do |event|  %>
        <%=  render event %>
<% end %>
<%== pagy_bootstrap_nav(@events_pagy) %>

# and then separately:
<% @upcoming.each do |upcoming|  %>
        <%=  render upcoming%>
<% end %>
<%== pagy_bootstrap_nav(@upcoming_pagy) %>
```

But these days, you could even wrap them in a turbo frame so that you can have completely independent contexts: https://turbo.hotwired.dev/handbook/frames

### Etiquette
I think i saw this same question both on stack overflow and on reddit? The general etiquette is to post all the links to a single source, otherwise x10 peopel might be answering the same question independently. 


I hope the above helps.

**Reply by ddnexus** [reply]
⚠️ You should also have different `:page_param`s or you will hit the same page in all objects. For example:

```ruby
@events_pagy, @events = pagy(events_scope) # use the default page_param: :page
@upcoming_pagy, @upcoming = pagy(upcoming, page_param: :upcoming_page)
```

**Reply by ogarocious** [reply]
Thank you so much for your help! After a few hours of tweaking I got a solution that is working however, the ransack search is not working with some records (before pagy it was working fine), meaning when I search an event of a particular event type it pulls in records that has and also doesn't have that event type

In my events controller I have:
```
  @q = Event.ransack(params[:q])
      @events = @q.result(distinct: true).active.order(event_start_date: :asc)
      @upcoming_q = Event.active.upcoming_events.order(event_start_date: :asc).ransack(params[:q])
      @upcoming_pagy, @upcoming_events = pagy(@upcoming_q.result(distinct: true), page_param: :upcoming_page, link_extra: 'data-turbo-frame="search"')
      
      @past_q = Event.active.past_events.order(event_start_date: :asc).ransack(params[:q])
      @past_pagy, @past_events = pagy(@past_q.result(distinct: true), page_param: :past_page, link_extra: 'data-turbo-frame="search"')
  end
```

In my events index ERB file I have:
```
<%= turbo_frame_tag 'search' do %>
         <h2>Upcoming Events</h2>
              <h6>Currently displaying <%= @upcoming_events.count %> of <%= pluralize(Event.upcoming_events.count, "upcoming event") %>.</h6> 
         <br>
         <%== pagy_bootstrap_nav(@upcoming_pagy) %>
         <br>
              <div class="event-list-wrapper">
                   <% @upcoming_events.each do |event| %>
                        <%= render 'event', event: event %>
                    <% end %>   
              </div> 
        
         <h3>Past Events</h3>
         <h6>Currently displaying <%= @past_events.count %> of <%= pluralize(Event.past_events.count, "past event") %>.</h6> 
         <%== pagy_bootstrap_nav(@past_pagy) %>
              <div class="event-list-wrapper past-event">
                  <% @events.past_events.each do |event| %>
                    <%= render 'event', event: event %>
                  <% end %>
              </div> 
         <% end %>
```

and my search form for Ransack is pointing to @q:
```
<%= search_form_for @q,  
    data: { controller: "places", action: "google-maps-callback@window->places#initMap" }, data: { turbo_frame: :results } do |f| %>
    <div class="formlook">
        <div class="event-search-container">
            <div class="event-name">
                <%= f.label "Event name", class: "form-label" %>
                <%= f.search_field :name_cont, placeholder: "Any event", class: "form-control" %>
            </div>
            <div class="geolocation">
                <%= f.label "Geolocation", class: "form-label" %>
                <%= f.search_field :address_or_city_or_state_or_country_or_continent_cont_all, class: "form-control", placeholder: "Search by continent, country, state, or city", data: { places_target: "field" } %>
            </div>
            <div class="dancestyle">
                <%= f.label "Dance style", class: "form-label" %>
                <%= f.select(:dance_styles_id_in, DanceStyle.all.sorted.pluck(:name, :id), { :include_blank => "All dance styles" }, { class: "form-select", id: "select-dancestyle", multiple: true, placeholder: "Any dance style" }) %>
            </div>
            <div class="eventtype">
                <%= f.label "Event Type", class: "form-label" %>
                <%= f.select :event_type_id_in, EventType.all.sorted.pluck(:name, :id), { :include_blank => "All event types" }, { id: "select-artist", multiple: true, placeholder: "Any event type", class: "form-select" } %>
            </div>
            <div class="month">
                <%= f.label "Month", class: "form-label" %>
                <%= f.select(:event_month_in, (Date::MONTHNAMES[1..12]), { prompt: "Any month" }, { id: "select-month", :include_blank => "Any month", class: "form-select", multiple: true }) %>
            </div>
            <div class="year">
                <%= f.label "Year", class: "form-label" %>
                <%= f.select :event_year_in, Date.today.year-2 .. Date.today.year+3, { prompt: "Any year" }, { id: "select-year", include_blank: true, class: "form-select", multiple: true }  %>
                <br>
            </div>
        </div>
     </div>
    <%= f.submit "Search!", class: "button-orange" %> <br><br>
    <%= link_to "Clear Search", request.path, class:"cancel-button" %>
<% end %>
```

Something is quite lining up with the @q search params and then my new @upcoming_q and @past_q, does anything look off?

Thanks in advance!

**Reply by ddnexus** [reply]
Despite the fact that "before pagy it was working", pagy has nothing to do with it anyway, because it paginates the scope that you feed to it.

The first thing I notice is that your displayed past events should be `@past_events` (filtered by ransack) instead of `@events.past_events` (not filtered)... 

I didn't check whether there is more mess somewhere else.

**Reply by benkoshy** [reply]
I concur with what @ddnexus. I cannot see this as a pagy issue. I would say it's more a rails / hotwire / ransack question.

However, I'll add me two cents now how to solve this: start simple, then add complexity.

- Remove turbo frames
- Remove ransack from this action.
- Remove pagy.
- Remove complex views.

Get it working in its most simplest form. When that's done, then add complexity: turbo_frames, pagy, THEN ransack, then complex views.

**Reply by ddnexus** [reply]
@ogarocious ?

**Comment by ddnexus** [comment, marked as answer]
Documentation: [Paginate multiple independent actions](https://ddnexus.github.io/pagy/docs/how-to/#paginate-multiple-independent-collections) and [Using different page_param](https://ddnexus.github.io/pagy/docs/how-to/#using-different-page_params)
